jcl/src/java.base/share/classes/java/lang/Class.java

```diff
@@ -1567,7 +1567,24 @@ Method getMethodHelper(
 			boolean scanInterfaces = false;
 			for (Method m: methods) {
 				MethodInfo mi = new MethodInfo(m);
-				myMethods.put(mi, mi);
+				MethodInfo prevMI = myMethods.put(mi, mi);
+				if ( prevMI != null) {
+					/* As per Java spec:
+					 * For methods with same signature (name, parameter types) and return type,
+					 * only the most specific method should be selected. 
+					 * Method N is more specific than M if:
+					 * N is declared by a class and M is declared by an interface; or
+					 * N and M are both declared by either classes or interfaces and N's
+					 * declaring type is the same as or a subtype of M's declaring type.
+					 */
+					Class<?> miDeclaringClz = mi.me.getDeclaringClass();
+					Class<?> prevMIDeclaringClz = prevMI.me.getDeclaringClass();
+					if ((miDeclaringClz.isInterface() && !prevMIDeclaringClz.isInterface())
+						|| (miDeclaringClz.isAssignableFrom(prevMIDeclaringClz))
+					) {
```

Comment:
Does the logic below to handle interface methods need similar validation before methods are added to `myMethods`?  line ~1603

Reply:
The methods in line `~1603` are from `getDeclaredMethods()` which doesn't include any parent interface methods. Hence the logic mentioned is not required.

Reply:
@DanHeidinga PR has been updated, please have another look.